src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -0,0 +1,288 @@
+/*-------------------------------------------------------------------------
+ *
+ * function_call_delegation.c
+ *    Planning logic for delegating a function call to a worker when the
+ *    function was distributed with a distribution argument and the worker
+ *    has metadata.
+ *
+ * Copyright (c), Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/pg_proc.h"
+#include "catalog/pg_type.h"
+#include "commands/defrem.h"
+#include "distributed/citus_custom_scan.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/commands.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/connection_management.h"
+#include "distributed/function_call_delegation.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/master_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/remote_commands.h"
+#include "distributed/shard_pruning.h"
+#include "distributed/version_compat.h"
+#include "distributed/worker_manager.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "optimizer/clauses.h"
+#include "parser/parse_coerce.h"
+#if PG_VERSION_NUM >= 120000
+#include "parser/parsetree.h"
+#endif
+#include "miscadmin.h"
+#include "tcop/dest.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+static bool contain_param_walker(Node *node, void *context);
+
+/*
+ * contain_param_walker scans node for Param nodes.
+ * returns whether any such nodes found.
+ */
+static bool
+contain_param_walker(Node *node, void *context)
+{
+	return IsA(node, Param);
+}
+
+
+/*
+ * TryToDelegateFunctionCall calls a function on the worker if possible.
+ * We only support delegating the SELECT func(...) form for distributed
+ * functions colocated by distributed tables, and not more complicated
+ * forms involving multiple function calls, FROM clauses, WHERE clauses,
+ * ... Those complex forms are handled in the coordinator.
+ */
+DistributedPlan *
+TryToDelegateFunctionCall(Query *query)
```

Comment:
Minor: We might consider splitting this function into two `ShouldDelegateFunctionCall() `, and `DelegateFunctionCall()` just for making this function smaller.

Reply:
Would probably cause a lot of duplicate work though (e.g. looking at metadata to get shard placements).